### PR TITLE
Feat: Implemented Data Loader for BioModels Simulations

### LINKED ISSUE #23: Implement Data Loader for Leaderboard Repository
This issue aims to implement a data loader for running BioModels simulations. The loader should accept a model name as input, automatically fetch required parameters from an existing YAML file, and execute the simulation. This will streamline model execution and remove the need for manual configuration.

Implement a data loader that:
- Accepts a BioModel name as input.
- Fetches simulation parameters (e.g., time steps, solver settings) from a predefined YAML file.
- Loads the corresponding SBML model dynamically.
- Executes the simulation with the extracted parameters.

Expected Outcome:
-  A fully automated BioModels data loader that eliminates manual parameter configuration.
-  Seamless execution of simulations based on predefined YAML inputs.
-  Improved scalability for running multiple models dynamically.

### PR DESCRIPTION
**Pull request by Rakesh-Seenu**
# For authors
image

## Description
This PR addresses issue #23  by implementing a fully functional and automated data loader that enables seamless execution of BioModels simulations using PyTorch Lightning.

Work done now: 
This contribution introduces a new SBMLDataModule that streamlines the workflow of running simulations by:

- Accepting a BioModel name as input to dynamically determine which simulation to run. (used hydra for connfiguration)
- Automatically fetching simulation parameters such as train_duration and test_duration from a pre-existing YAML configuration file.
- Loading and executing SBML models using COPASI through the basico Python package.
- Returning PyTorch-compatible Datasets via an iterable SBMLTimeCourseDataset class.
- Providing both train_dataloader() and test_dataloader() for modular use in training pipelines.

Key Features:
-  Modular and extensible design via LightningDataModule
-  Built-in error handling for missing files or misconfigured YAML
-  Dynamic model path resolution using Hydra
-  Fully reproducible SBML simulation runs for both training and testing stages

## Additional resources added as  in picture above
Developer Documentation:
- Overview of module structure and function responsibilities.
- Notes on simulation execution and dataset generation.

DevOps / CodeOps Docs:
- Instructions for setting up local environments.
- Dependency management and config file layout.

Usage Tutorial:
Step-by-step walkthrough showing how to:
- Add new SBML models and YAML files
- Run the SBMLDataModule in a training loop
- Extend the dataset for custom experiments

Next PR:
- Enhance the implementation of the iterable dataset
- implement the batch size for training in future
- right now i have commented out the validation dataloader in next pr it will be implemented

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [ ] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

**Comment by Rakesh-Seenu** [comment]
@gurdeep330 Please let me know any changes is required. Thank you.

**Comment by Rakesh-Seenu** [comment]
> @Rakesh-Seenu code coverage failed. Could you please look into it? Thx

sure, I will look into it as soon as reach home.

File: docs/ops/CodeOps.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+This guide will cover the following topics relevant for our CodeOps:.
+
+🗸 **GitHub repo organization**: About the organization of the repository on GitHub.<br>
+🗸 **GitHub templates**: Explains the predefined templates that are provided in a standardized format to provide details on proposed changes, new features, or bugs.<br>
+🗸 **Coding Practices**: Outlines the coding standards that we strive to follow to ensure the quality, maintainability, and consistency of our codebase.<br>
+🗸 **Testing locally**: Steps to run unit tests locally before committing your code, thereby reducing workflow failures on GitHub Actions prior to commit.<br>
+🗸 **Keywords in Commit Messages**: Specific keywords in commit messages that can trigger release on GitHub Actions.<br>
+🗸 **Resources for Further Reading**: Additional reading material.<br>
+
+## GitHub Repo Organization
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply docs/ops/CodeOps.md:38]: No the folder structure is not of this repo, I will remove it and add links to the code and dev ops of the AIAgents4Pharma.

File: mkdocs.yml [hunk 2]
```diff
@@ -6,11 +6,21 @@ nav:
 - Getting started: user-guide/index.md
 - Data: 
   - Data: data/index.md
+  - Tutorials: 
+    - SBML Dataloader: notebooks/sbml_dataloader.ipynb
   - Models: 
     - System Bio Model: data/sys_bio_model.md
     - Basico Model: data/basico_model.md
+    - Sbml Dataloader: data/sbml_dataloader.md
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply mkdocs.yml:14]: I will update as suggested.

File: vpeleaderboard/data/configs/BIOMD0000000537_url.yaml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+train_duration: 100
+# val_duration: 30
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply vpeleaderboard/data/configs/BIOMD0000000537_url.yaml:2]: There was no specific reason for including only the training and testing stages; they were added as initial references. I plan to extend the implementation by incorporating a validation stage as well.

File: vpeleaderboard/data/src/basico_model.py [hunk 4]
```diff
@@ -39,10 +39,10 @@ def validate_sbml_file_path(self):
         if not self.sbml_file_path:
             print("100")
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply vpeleaderboard/data/src/basico_model.py:40]: No, just to test I added the print statement.

File: vpeleaderboard/data/src/sbml_dataloader.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        Returns:
+            DataLoader: The DataLoader for the training dataset.
+        """
+        train_df = basico.run_time_course(
+            duration=self.config['train_duration'],
+            use_initial_values=False
+        )
+        dataset = self.SBMLTimeCourseDataset(train_df)
+        return DataLoader(dataset)
+    # def val_dataloader(self) -> DataLoader:
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply vpeleaderboard/data/src/sbml_dataloader.py:141]: There was no specific reason for not including validation data loader. I plan to extend the implementation by incorporating a validation stage as well.

File: vpeleaderboard/data/src/sbml_dataloader.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        Creates the DataLoader for the training dataset based on the SBML simulation results.
+
+        Returns:
+            DataLoader: The DataLoader for the training dataset.
+        """
+        train_df = basico.run_time_course(
+            duration=self.config['train_duration'],
+            use_initial_values=False
+        )
+        dataset = self.SBMLTimeCourseDataset(train_df)
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply vpeleaderboard/data/src/sbml_dataloader.py:139]: I will create a separate issue for the above feature request and implement it in a subsequent pull request.

File: vpeleaderboard/data/src/sbml_dataloader.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            FileNotFoundError: If the SBML model or YAML config file is not found.
+            ValueError: If the YAML config file is empty or missing required keys.
+        """
+        script_dir = os.path.dirname(__file__)
+        script_dir = "\\".join(script_dir.split("\\")[:-1])
+        with hydra.initialize(version_base=None,
+                              config_path="../configs"):
+            cfg = hydra.compose(config_name="config")
+
+        model_directory = os.path.join(script_dir, cfg.model_directory)
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply vpeleaderboard/data/src/sbml_dataloader.py:79]: The model can directly be loaded from the config, but when we try to use the file as a package, we'll face file not found error as the base directory would be different.

File: vpeleaderboard/data/configs/BIOMD0000000064_url.yaml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+train_duration: 6
+test_duration: 5
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply vpeleaderboard/data/configs/BIOMD0000000064_url.yaml:2]: You're absolutely right — extending the configuration to include detailed simulation and post-processing parameters will significantly improve the flexibility and reproducibility of the training and test data generation process. For this initial implementation, I’ve kept the configuration minimal to focus on the integration pipeline and ensure everything works end-to-end.

I’ll create a separate issue to track this enhancement and align it with the broader data generation goals you outlined.

**Comment by Rakesh-Seenu** [comment]
@dmccloskey I just wanted to let you know that I've opened an issue regarding the macOS-specific behavior we're encountering with the current PR. Lilija is reaching out to the authors of basico to clarify the situation. For now, we will be ignoring the coverage tests on macOS as we work through the issue.

**Review by dmccloskey — APPROVED** [review]
Apologies for the delay in review. I somehow missed the re-request. The PR looks good to me. There are a few line comments that I think could be removed and a typo that I suggest a change for.

**Comment by Rakesh-Seenu** [comment]
[vpeleaderboard/data/src/sbml_dataloader.py](https://github.com/VirtualPatientEngine/VPELeaderboard/pull/25/files/850e7acb519e7d8e7851f2b3b7530b7d2460a9d9#diff-8b5fed8dc9dd69a3764dce35273319160756e11402c50827939cb3636f1de31c)
                              config_path="../configs"):
            cfg = hydra.compose(config_name="config")

        model_directory = os.path.join(script_dir, cfg.model_directory)
Member
@[dmccloskey](https://github.com/dmccloskey) dmccloskey [5 days ago](https://github.com/VirtualPatientEngine/VPELeaderboard/pull/25#discussion_r2047451253)
Gotcha. But I think now the commented lines can be removed?

@dmccloskey No, we need it i have uncommented it.

File: vpeleaderboard/data/src/basico_model.py [hunk 9]
```diff
@@ -1,5 +1,5 @@
 """
-BasicoModel class for loading SBML models
+BasicModel class for loading SBML models
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply vpeleaderboard/data/src/basico_model.py]: I have updated it as suggested.

File: tests/app/data/test_markdown.py [hunk 10]
```diff
@@ -12,14 +12,31 @@ def test_validate_sbml_file_path_success():
...
     assert len([f for f in os.listdir("vpeleaderboard/data/models") if f.endswith(".xml")]) > 0
-    model = BasicoModel(sbml_file_path="vpeleaderboard/data/models/BIOMD0000000064_url 1.xml")
+
+    path = "vpeleaderboard/data/models/BIOMD0000000064_url.xml"
+    sbml_file_path = os.path.abspath(path)
+
+    model = BasicoModel(sbml_file_path=sbml_file_path)
     assert model is not None
 
+# def test_get_model_metadata():
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply tests/app/data/test_markdown.py]: No, we don’t need that function as we already have another one that tests get_model_metadata(). I’ve removed the unnecessary one to streamline the process.

**Review by dmccloskey — APPROVED** [review]

File: .github/workflows/tests.yml [hunk 11]
```diff
@@ -50,7 +50,7 @@ jobs:
 
     strategy:
         matrix:
-          os: [ubuntu-latest, macos-13]
+          os: [ubuntu-latest, macos-13]
```
Comment (gurdeep330) [inline .github/workflows/tests.yml:53]: Now that the basico error is resolved, I'd recommend changing `macos-13` to `macos-latest`

File: tests/app/data/test_markdown.py
Comment (gurdeep330) [inline tests/app/data/test_markdown.py:1]: I'd suggest creating a global variable at the top of the script to store the path to the model:
`path = "vpeleaderboard/data/models/BIOMD0000000064_url.xml"`

Use the global variable wherever the path is needed.

File: tests/vpeleaderboard/data/test_basico_model.py
Comment (gurdeep330) [inline tests/vpeleaderboard/data/test_basico_model.py:1]: I'd suggest creating a global variable at the top of the script to store the path to the model:
`path = "vpeleaderboard/data/models/BIOMD0000000064_url.xml"`

Use the global variable wherever the path is needed.

File: tests/vpeleaderboard/data/test_sbml_dataloader.py
Comment (gurdeep330) [inline tests/vpeleaderboard/data/test_sbml_dataloader.py:1]: I'd suggest creating a global variable at the top of the script to store the path to the model:
`path = "vpeleaderboard/data/models/BIOMD0000000537_url.xml"`

Use the global variable wherever the path is needed.

File: tests/vpeleaderboard/data/test_sys_bio_model.py
Comment (gurdeep330) [inline tests/vpeleaderboard/data/test_sys_bio_model.py:1]: I'd suggest creating a global variable at the top of the script to store the path to the model:
`path = "vpeleaderboard/data/models/BIOMD0000000064_url.xml"`

Use the global variable wherever the path is needed.

**Review by gurdeep330 — CHANGES_REQUESTED** [review]
Nice work 👏 
I have only very minor comments.

File: .github/workflows/tests.yml (hunk 11 above)
Comment (Rakesh-Seenu) [reply .github/workflows/tests.yml:53]: Okay, I will update it

**Comment by Rakesh-Seenu** [comment]
@gurdeep330 Thank you for the suggestion! I have implemented the global variable to store the model path at the top of the script, and I've used it wherever the path is needed in all the mentioned files.
